hazelcast-jet-hadoop/src/main/java/com/hazelcast/jet/hadoop/impl/ReadHdfsNewApiP.java

```diff
@@ -0,0 +1,255 @@
+import org.apache.hadoop.mapreduce.Job;
+import org.apache.hadoop.mapreduce.MRJobConfig;
+import org.apache.hadoop.mapreduce.RecordReader;
+import org.apache.hadoop.mapreduce.TaskAttemptID;
+import org.apache.hadoop.mapreduce.lib.input.TextInputFormat;
+import org.apache.hadoop.mapreduce.task.TaskAttemptContextImpl;
+import org.apache.hadoop.util.ReflectionUtils;
+
+import javax.annotation.Nonnull;
+import java.io.ByteArrayOutputStream;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+import static com.hazelcast.jet.Traversers.traverseIterable;
+import static com.hazelcast.jet.impl.util.ExceptionUtil.sneakyThrow;
+import static java.util.Collections.emptyList;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * See {@link HdfsSources#hdfs}.
+ */
+public final class ReadHdfsNewApiP<K, V, R> extends AbstractProcessor {
+
+    private final Configuration configuration;
+    private final InputFormat inputFormat;
+    private final Traverser<R> trav;
+    private final BiFunctionEx<K, V, R> projectionFn;
+    private final BetterByteArrayOutputStream cloneBuffer = new BetterByteArrayOutputStream();
+    private final DataOutputStream cloneBufferDataOutput = new DataOutputStream(cloneBuffer);
+
+    private ReadHdfsNewApiP(
+            @Nonnull Configuration configuration,
+            @Nonnull InputFormat inputFormat,
+            @Nonnull List<InputSplit> splits,
+            @Nonnull BiFunctionEx<K, V, R> projectionFn
+    ) {
+        this.configuration = configuration;
+        this.inputFormat = inputFormat;
+        this.trav = traverseIterable(splits)
+                .flatMap(this::traverseSplit);
+        this.projectionFn = projectionFn;
+    }
+
+    @Override
+    public boolean isCooperative() {
+        return false;
+    }
+
+    @Override
+    public boolean complete() {
+        return emitFromTraverser(trav);
+    }
+
+    @SuppressWarnings("unchecked")
+    private Traverser<R> traverseSplit(InputSplit split) {
+        RecordReader<K, V> reader;
+        try {
+            TaskAttemptContextImpl attemptContext = new TaskAttemptContextImpl(configuration, new TaskAttemptID());
+            reader = inputFormat.createRecordReader(split, attemptContext);
+            reader.initialize(split, attemptContext);
+        } catch (IOException | InterruptedException e) {
+            throw sneakyThrow(e);
+        }
+
+        return () -> {
+            try {
+                while (reader.nextKeyValue()) {
+                    // we have to clone the the value returned by `reader.getCurrentKey()` because
+                    // the next call to `reader.nextKeyValue()` will mutate the returned value instead
+                    // of creating a new value. Same for value.
+                    K key = serdeWritable(reader.getCurrentKey());
```

Comment:
this seems like a really bad idea and will be very slow. isn't there some other way to do it? In old API you could use .createKey(). Each Reader should have its own createKey/Value implementation.

Reply:
I had another look and seems this method is removed in new api. There's another method called `ReflectionUtils.cloneWritableInto`, but still it will be slow to serialize/deserialize each key/value.

Reply:
I had one more look, and actually this is completely wrong because there is no guarantee that the items are Writable. When using Avro/Parquet then this is not the case.